### DISCUSSION TITLE
Parallel Execution of Independent Nodes

### DISCUSSION THREAD

**Discussion by torealise** [comfyanonymous/ComfyUI#3683] [Ideas]
Can we implement parallel execution of independent nodes in ComfyUI to improve performance?

**Description:**

Allow nodes that do not depend on each other to run simultaneously. This can optimize resource usage and reduce processing time.

**Examples:**

Image Processing:

- Node A: Loads Image 1, preprocesses it, and sends it to ControlNet.
- Node B: Loads Image 2 and sends it to IPAdapter.
- Node C: Segments Image 3.

All these operations can run in parallel.

**Benefits:**

- Faster execution.
- Better use of CPU/GPU resources.
- More efficient handling of complex workflows.

If determining which nodes are independent is complex, this configuration could be set manually for each workflow.

Would it be possible to add this feature?

**Comment by ltdrdata** [comment]
This requires an overhaul of the execution model and in-depth consideration of various corner cases.

For example, tasks that require VRAM should not operate in parallel. Otherwise, you'll encounter the nightmare of VRAM shortage. There are complex issues such as policy decisions on which tasks should run in parallel and which should not, and how to ensure that each node developer correctly applies these policies without errors.

To implement this, it is not a matter that can be addressed simply in an issue. It will need to be refined through lengthy discussions in a discussion forum.

**Comment by torealise** [comment]
> This requires an overhaul of the execution model and in-depth consideration of various corner cases.
> 
> For example, tasks that require VRAM should not operate in parallel. Otherwise, you'll encounter the nightmare of VRAM shortage. There are complex issues such as policy decisions on which tasks should run in parallel and which should not, and how to ensure that each node developer correctly applies these policies without errors.
> 
> To implement this, it is not a matter that can be addressed simply in an issue. It will need to be refined through lengthy discussions in a discussion forum.

Could we create a tool that allows users to manually specify which nodes can run in parallel for each specific workflow?

This would help calculate the VRAM load for each case and use memory more efficiently. This way, we can avoid VRAM shortages and handle complex workflows more effectively.

By letting users decide which tasks run in parallel, we solve the problems with automatic parallelization and give them control over resource use.

**Reply by identitycrisis2077** [reply]
Can we first go through all the concerns we have with making this entirely parallel and backtrack from there? I would go with a parallel first approach and then jot down and discuss how to solve for the critical sections where we need sequential access. One thing which comes to my mind is Jenkins and it's milestones. My initial hunch or thought is for users / node developers to call out the critical resource their node will be blocking. Thinking of something like KSampler could acquire like a lock on gpu so until the next node which acquires a gpu lock, we can continue. The node graph already describes which nodes can be parallel and sequential.

**Reply by identitycrisis2077** [reply]
### Pros

* Easier to determine the nodes which can block critical section
* Easier to implement as well 

### Cons
*  Most of the nodes are VRAM / GPU intensive serving as a bottleneck? 
* One size fits all is not optimal. Some nodes which require VRAM might call a lock on the vram but still might have enough vram left for other operations
* Not seamless for the user. They need manual configuration
* The user needs to understand how parallelism and critical section works though it's a no-code solution for writing parallel computation.

**Reply by identitycrisis2077** [reply]
Another option is to dynamically generate parallel execution. The first execution of a workflow would be sequential execution but the nodes will collect it's corresponding footprint and we can make an opinionated guess on which nodes can be parallelized based on the footprint and available resource. If it's not available, the node will queue and wait on it's execution until required resources are freed. Might not be optimal but still better than 100% sequential.

**Reply by WASasquatch** [reply]
This presents a great opportunity to have GPU node flags (useful information to have available about nodes in any case), and subsequently, add a dynamic GPU selection dropdown on these flagged nodes, which is passed to the back-end determining the device the operation runs on, so when you have a 6 card 4090 bench, you can do 6 ksamplers in parallel. We should start thinking beyond the novelty hobbyists using ComfyUI. 

And hey, if you run 3 jobs on one device that is GPU intensive, maybe, like, don't do that. That's on them and probably shouldn't be a limiting factor in mature features already enjoyed in other software without really making the risk an issue. They assume people know what they're doing when they manually enable/use a feature.

**Reply by ltdrdata** [reply]
IMO, a GPU pool approach seems better than critical sections. 
While nodes are executing in parallel, nodes that require important resources like GPU can request a resource object in a way similar to requesting a unique_id in hidden parameters. 
Until the node's execution is completed, the usage rights for that GPU device are allocated only to that specific node.

In more detail, as @WASasquatch  suggested, we could expose it as a 'required' parameter, allowing users to select the GPU device. In this case, during parallel execution of the workflow, the node would remain in a queue state until the specified GPU device becomes available.

Or it would be possible for a single node to request multiple GPU devices.

```
  {
      'required': { 
                "device": ("DEVICE", { 'is_multiple': False } ) 
      },
  }
```

```
  {
      'required': { 
                "device": ("DEVICE", { 'is_multiple':True, 'count': 'all' } ) 
      },
  }
```

**Reply by WASasquatch** [reply]
I like the idea of input param. With new UI going on could be expanded with a device filter. Say you want X nodes to only use devices `[1, 2]` because 0 is some low class card that you know won't handle a specific job, or maybe used elsewhere

_Looking at my friend with his 1080 and 3080 and shenanigans he goes through with ML lmao_

**Reply by identitycrisis2077** [reply]
Just humor me, wouldn't running multiple instances of core workflow achieve the same as running on a GPU pool? Allowing dynamically scaling a workflow to run multiple requests concurrently would allow for cost effective scaling of workflows. This is not something which a hobbyist might require but this has potential on the cloud.

**Comment by pulkit-jain** [comment]
I think this is a very much required feature otherwise ComfyUI will not harnessing it full potential.

There are multiple use cases for this!

Better is to have a thread node/group to denote the nodes can be executed in threads and then these nodes can run parallely.

**Comment by ltdrdata** [comment]
@guill Previously, it was somewhat premature to discuss the parallelization of the execution model, but now that #2666 has been merged, it seems possible to discuss the next steps towards parallelization.

Are you perhaps interested in this topic as well?

**Reply by guill** [reply]
Definitely! I started doing some research on this (and brought it up a few times in Discord/Matrix) as #2666 was nearing completion.

(Apologies in advance for the wall of text. This ended up being longer than I intended😅)

The execution model is already most of the way there for supporting parallel execution. When we try to execute a node, the internal `execute` function can return one of three `ExecutionResult` results:
1. `SUCCESS` - It finished and has results.
2. `FAILURE` - It failed with an error.
3. `PENDING` - It's waiting on something and you should come back to it later.

Currently, `PENDING` is used for the execution of dynamically generated subgraphs, but it could be used just as easily for an asynchronous function.

While supporting standard Python coroutine `async` semantics should be pretty easy (making a proof-of-concept is likely a weekend project for someone with asyncio experience), that doesn't really solve the original request. As far as I could find, there isn't an easy way to perform PyTorch operations in an asynchronous way and determine when they're complete.

Once we get into needing actual threads, things get complicated. Because PyTorch operations really are asynchronous from the CPU perspective (and only force synchronization once you need results back on the CPU), it doesn't look like things are quite as easy as spinning up a new thread for each node and just waiting until it's complete. Technically, when a node "completes" execution from Python's perspective, the PyTorch operations performed in it may still be occurring. In addition to worrying about normal threading issues, we need to make sure things are in line with PyTorch's internal synchronization model as well.

The other big issue is the one about resource contention discussed above. Ideally, I'd like to avoid making workflow creators worry about things like that. If a workflow that works on 12GB of VRAM needs to be modified to take advantage of a machine with 24GB of VRAM, that's going to make sharing (and developing) workflows a lot more difficult. The direction I was leaning for solving this was:

1. Create the concept of quantified resources. The default ComfyUI would pre-fill this with some resources, and custom nodes could add others. The state would look something like:
```python
{
    "RAM": 68719476736,
    "device1.VRAM": 34359738368,
    "BlenderInstances": 1,
    "Webcam": 1,
    # ...
}
```
2. Prior to calling the execution function on a node, call a new function (with all the evaluated inputs) that just has the job of estimating resource usage (e.g. by looking at tensor shape):
```python
def get_required_resources(self, image, scale_x, scale_y):
    return {
        f"{image.device}.VRAM": image.element_size() * image.numel() * scale_x * scale_y,
    }
```
3. Only actually call the execution function on nodes where resource usage is sufficient.

Theoretically, that would allow the same workflow to run on two very different machines with the low-spec machine running one node at a time while the powerhouse runs as many as possible. Obviously there's some extra work there for node authors, but it'll be easier to experiment with different methods of estimating those resources for nodes that don't explicitly define them  (as described earlier in this discussion) once the support exists in the core engine.

**Reply by ltdrdata** [reply]
Oh, what you're talking about is more about distributed computing rather than supporting multi-GPU setups on a single machine.

If parallel execution is supported, it might be a good idea to discuss that as well.

**Reply by guill** [reply]
I definitely think distributed computing is worth talking about at some point, but I'm really just thinking about one machine at the moment. The goal of the quantified resources isn't to find the right machine to run the node on -- just to make sure we aren't running too many nodes at the same time.

Even if the only use-case we're trying to solve is running one node on each of two GPUs concurrently, the issues I mentioned will still be relevant. Unless we're going to rely on the user to guarantee that all inputs are already on the correct GPU (which may be difficult in cases where the tensor is contained within a dictionary like `CONDITIONING`), we're still going to have to worry about concurrent threads interacting with tensors on a single GPU.

**Reply by WASasquatch** [reply]
This is why I was talking about flagging nodes as GPU, and by doing so, similar to "control_after_generate" you'd have a device select listing devices. Anything hitting this node should run through and define devices based on selection. If the same device is encountered you wait, else run on that GPU. Id imagine there would be edge cases all over. For example, if the results of all are fed to vae decode, and subsequently batched together, I imagine those batch nodes would need to weight until both inputs are satisfied before moving to the next batch node, etc.

**Reply by aikitoria** [reply]
Could automatically managing resources from a pool perhaps be done as a second step? My assumption here is that people with access to multi GPU systems that can take advantage of this will have enough experience to modify their workflows accordingly, so it'd be a great first step to get something going at all.

For example, how about flagging each GPU node with a GPU index to run on. During execution it then schedules all possible nodes in parallel, with the restriction that two nodes assigned to the same GPU cannot run at the same time. This would let us easily build workflows that process batches in parallel on multiple GPUs by copy pasting a few nodes, and then join the results back together to finalize.

When designing a batching workflow for multiple GPUs it doesn't really make sense to me to always want it shareable with systems that have less GPUs. For example I might have loaded multiple models on specific GPUs to create some kind of pipeline with the assumption that I have this particular amount of VRAM available to do it.

An example workflow might be loading the text encoder and vae for flux on GPU 0, preparing the prompts there, then generating 4 batches in parallel on GPUs 1-4, decoding all the images back on GPU 0, and finally upscaling them on GPU 5.

If we could go one step further and allow partially overlapping multiple instances of such workflow, with the same restriction on parallel processing on the same GPU, so that the flux for prompt 2 can already run in parallel with the upscale for prompt 1, that would be truly amazing.

Basically this would avoid the whole resource contention problem by creating multiple resource pools (one per GPU) that work similar to the current single-GPU implementation and letting the user specify which to use per node.

Once that system exists, more fancy pooling could be built on top, for example with a node that runs a subgraph as many times as you have GPUs, or similar.

**Reply by WASasquatch** [reply]
> Could automatically managing resources from a pool perhaps be done as a second step? My assumption here is that people with access to multi GPU systems that can take advantage of this will have enough experience to modify their workflows accordingly, so it'd be a great first step to get something going at all.
> 
> For example, how about flagging each GPU node with a GPU index to run on. During execution it then schedules all possible nodes in parallel, with the restriction that two nodes assigned to the same GPU cannot run at the same time. This would let us easily build workflows that process batches in parallel on multiple GPUs by copy pasting a few nodes, and then join the results back together to finalize.
> 
> When designing a batching workflow for multiple GPUs it doesn't really make sense to me to always want it shareable with systems that have less GPUs. For example I might have loaded multiple models on specific GPUs to create some kind of pipeline with the assumption that I have this particular amount of VRAM available to do it.
> 
> An example workflow might be loading the text encoder and vae for flux on GPU 0, preparing the prompts there, then generating 4 batches in parallel on GPUs 1-4, decoding all the images back on GPU 0, and finally upscaling them on GPU 5.
> 
> If we could go one step further and allow partially overlapping multiple instances of such workflow, with the same restriction on parallel processing on the same GPU, so that the flux for prompt 2 can already run in parallel with the upscale for prompt 1, that would be truly amazing.
> 
> Basically this would avoid the whole resource contention problem by creating multiple resource pools (one per GPU) that work similar to the current single-GPU implementation and letting the user specify which to use per node.
> 
> Once that system exists, more fancy pooling could be built on top, for example with a node that runs a subgraph as many times as you have GPUs, or similar.

Yeah that's what I am saying about GPUs. In general having ability to choose device a GPU node runs on is ideal. For example with eGPU and my older laptop (my new one has 4090 💪) I can't use it cause it defaults to cuda device:0. Though I didn't look into the problem much it goes to show even a modest setup could benefit let alone a bench

**Reply by vvauijij** [reply]
> I definitely think distributed computing is worth talking about at some point, but I'm really just thinking about one machine at the moment.

I was wondering if production systems using ComfyUI have already faced the need for distributed computing? I’d like your feedback on adding a general-purpose RPC node to execute remote nodes (and a tool to wrap and deploy nodes as services outside ComfyUI). This would allow asynchrony to be encapsulated in network calls without considering PyTorch's internal synchronization model.

**Reply by aikitoria** [reply]
I'd like a low-overhead generic RPC node. For example, generating images/latents on one system and sending them to another for second stage processing without having to convert it to a png in the middle or whatever. We can assume extremely fast networking between systems in a cluster so such encoding would be pure unnecessary overhead. https://github.com/nux1111/ComfyUI_NetDist_Plus already tries to add RPC nodes but they are difficult to use since they try to be too clever, can't handle multi-stage workflows, and add huge needless overhead for serialization (why does it not just send the raw latents/images...)

While not 100% optimal, this would get us a lot closer to solving the parallel execution in one workflow problem (just send out a bunch of RPCs to more Comfy instances on the other GPUs).

I'm not sure I understand the deployment you are suggesting. Couldn't the RPC just consist of executing a Comfy API workflow on another instance, so we do not need to deploy anything specific, as the "main" workflow can configure the other instances? Only question would be how to retrieve the results with high efficiency (should *not* involve saving files to disk, encoding images, converting binary data to base64 for no reason, additional network calls, or anything similarly silly)

**Reply by vvauijij** [reply]
Thanks for sharing feedback and highlighting the limitations and overheads of existing RPC-nodes! 

> I'm not sure I understand the deployment you are suggesting. Couldn't the RPC just consist of executing a Comfy API workflow on another instance, so we do not need to deploy anything specific, as the "main" workflow can configure the other instances? 

Deploying a single node would allow to significantly optimize model loading into RAM (as mentioned in other discussions, there are many problems with model caching). As of now, running remote Comfy instance seems to be more straightforward and useful

> Only question would be how to retrieve the results with high efficiency (should not involve saving files to disk, encoding images, converting binary data to base64 for no reason, additional network calls, or anything similarly silly)

I haven't explored https://github.com/comfyanonymous/ComfyUI/pull/2666 in details yet, but as mentioned above, PENDING status may be used to implement asynchronous retrieval of RPC-call results

**Reply by WASasquatch** [reply]
> I'd like a low-overhead generic RPC node. For example, generating images/latents on one system and sending them to another for second stage processing without having to convert it to a png in the middle or whatever. We can assume extremely fast networking between systems in a cluster so such encoding would be pure unnecessary overhead. https://github.com/nux1111/ComfyUI_NetDist_Plus already tries to add RPC nodes but they are difficult to use since they try to be too clever, can't handle multi-stage workflows, and add huge needless overhead for serialization (why does it not just send the raw latents/images...)
> 
> While not 100% optimal, this would get us a lot closer to solving the parallel execution in one workflow problem (just send out a bunch of RPCs to more Comfy instances on the other GPUs).
> 
> I'm not sure I understand the deployment you are suggesting. Couldn't the RPC just consist of executing a Comfy API workflow on another instance, so we do not need to deploy anything specific, as the "main" workflow can configure the other instances? Only question would be how to retrieve the results with high efficiency (should *not* involve saving files to disk, encoding images, converting binary data to base64 for no reason, additional network calls, or anything similarly silly)

I believe you can loop over the tensor keys and make a dictionary with tensor value with the `.tolist()` as then you can serialize this and send it with API calls.

**Comment by torealise** [comment]
Here’s a concept I’d like your feedback on:

 1. Users design and validate their workflow in ComfyUI.
 2. Once confirmed, they activate a profiler that tracks each node’s processing time, VRAM usage, and resource consumption.
 3. The profiler generates a resource map detailing system performance.
 4. A script analyzes this map, optimizing the workflow specifically for the hardware in use.
 5. The optimized settings are saved for future runs, maximizing efficiency on that system.

**Reply by aikitoria** [reply]
This sounds way too complicated. Just let the user configure how it should be distributed.

**Comment by erikturn** [comment]
While the features to control where different parts of a job are executed, and/or split models across GPUs are necessary in the long term, in the near term the focus should be on adding support for the queue to execute jobs against all available local GPUs (or a subset) and a list of remote server instances.   Should be relatively simple to add support for concurrent jobs at the queueing point and allow the user to select which compute resources are available for queueing jobs. If I have several mixed GPUs locally, I may want to omit the one with less VRAM if I am running bigger models in the batch, or manually add a list of remote server addresses.  Also need a mechanism for the remote jobs to return the output for the local pipeline to display and/or write to disk.    Once queueing takes advantage of distributing the workload, then adding more granular controls for which parts of the pipeline run on which GPU or remote server instance or tagging them for parallelism dependencies makes a lot of sense, it's just a lot more complicated and should be done after basic job queueing is working.

It's pretty easy to get multiple local GPUs or remote server instances going... older gpus with decent VRAM are starting to get pretty cheap, PCIE riser solutions and mining rig power supply setups let you easily add ~28 gpus to a single system, so support for multigpu stuff is going to be increasingly more expected and useful in the future.  e.g. I have a server with 8xM40x24gb's, one with 6x1060x6gb's, and have another one under assembly with 12xK80xdual12gb+2xM40x24gb cards, all of which was pretty cheap to pick up on ebay for less than the cost of the (better, but expensive) current generation cards, and when prices on RTX cards start dropping off after the nvidia 5000 series ships, there will be more better gpus flooding the channels....

As such I think the priority should be to get the queueing to take advantage of scheduling jobs to multiple GPUS first....

**Reply by aikitoria** [reply]
This is something you can already do with tools like SwarmUI (running one Comfy instance per GPU and forwarding jobs to them), while for running parts of a workflow in parallel there is no good solution at all yet.

**Reply by ltdrdata** [reply]
Although SwarmUI does provide such a scenario, it still needs improvement since its usability is completely different from being able to control it at the workflow level in ComfyUI.

Among the currently available methods, this approach using multiple instances is one way to achieve parallel processing.
https://github.com/city96/ComfyUI_NetDist

**Reply by aikitoria** [reply]
I've been using https://github.com/nux1111/ComfyUI_NetDist_Plus, works great on my machine with 8 GPUs to generate basic images in parallel and join them all back in a batch for the UI to display. But I haven't yet figured out how to create proper multi-stage distributed workflows with it (think jailing t5 and and sd 1.5 to one GPU, then generating conditioning for all on that one, generating the images in parallel on multiple GPUs, joining those back together in a batch, then upscaling on the separate GPU again). Or any other workflow that would require multiple forks and joins. Maybe I'm just not using it right.

**Reply by asenasen123** [reply]
> I've been using https://github.com/nux1111/ComfyUI_NetDist_Plus, works great on my machine with 8 GPUs to generate basic images in parallel and join them all back in a batch for the UI to display. But I haven't yet figured out how to create proper multi-stage distributed workflows with it (think jailing t5 and and sd 1.5 to one GPU, then generating conditioning for all on that one, generating the images in parallel on multiple GPUs, joining those back together in a batch, then upscaling on the separate GPU again). Or any other workflow that would require multiple forks and joins. Maybe I'm just not using it right.

I execute the following instructions：
 CUDA_VISIBLE_DEVICES=0 python main.py --listen --port 8188 
 CUDA_VISIBLE_DEVICES=1  python main.py --listen --port 8288 
 
 Why is the following error displayed:
 503 Server Error: Service Unavailable for url: http://127.0.0.1:8288/queue
What is the cause of the error? How to solve it？

**Comment by erikturn** [comment]
Yeah, I have StableSwarmUI running and it cranks up 8 instances for the GPUs easily enough..... for some reason I can't install NetDistPlus as it fails, but I do have NetDist sorta working after looking at the examples more closely... it hangs if I include the GPU 0 server in the workspace, maybe because I'm remote desktop'd in to the ubuntu host and that's using GPU0 to accelerate.  Anyway, it works ish, but it's still pretty clunky.  for example, it needs to finish all images in the current set before it can move on to the next queued job item... in my case some GPUs are on full x16 bandwidth riser cables and some are on 1x extenders, so the gpus idle out while waiting on the ones with the slower links to finish transferring data. Also, it's super irritating that the remote server ports change when they crash or get restarted.  the ports are all under python so the original ports don't get freed up unless you reboot or restart python. It also craps out if any one of the remote hosts aren't available. So... not very robust...

It works OK, but long term I'd rather see multigpu/remotehost support in the queuing directly and more robustly.... when designing things to be able to split independent nodes across multiple gpus and manage dependencies, the work to distribute work items from the queue across local and remote GPUs should also get done.. it's all interrelated and there are valid reasons to want to split a single job across multiple GPUs versus wanting to batch multiple jobs across a GPU pool.  Fundamentally it needs to be able to manage assigning various workstreams to the available local and remote GPUs, and managing ordering dependencies, whether they are whole jobs or subcomponents, and manage caching/reuse in gpu memory, ....it's kind of all the same thing if everything becomes work items that need to be scheduled on available GPUs, and there are different load balancing strategies one might want depending on how the cluster is being used.  (e.g. for cranking out images in batch vs accelerating a single workflow you would split things across gpus differently with maximizing  load efficiency by having the main model cached on all the GPUS vs maximizing parallelism for a single job by dedicating gpus to specific models)

**Comment by fedor-evolox** [comment]
What about the [MultiGPU](https://github.com/neuratech-ai/ComfyUI-MultiGPU) custom node?

**Reply by aikitoria** [reply]
This node is used to offload weights to specific GPUs in case you do not have enough VRAM all on one, it does not allow parallel execution.

**Reply by fedor-evolox** [reply]
> This node is used to offload weights to specific GPUs in case you do not have enough VRAM all on one, it does not allow parallel execution.

Well, it is also helpful in some cases.

**Comment by enn-nafnlaus** [comment]
Honestly, I'd be happy if it could even just load multiple files at once, or load and save two files at once... let alone divvy out GPU tasks.

**Reply by Bratzmeister** [reply]
It'd be great if you could have CLIP and VAE loaded on one gpu and the unet on another card all in the same comfyui instance. Doesn't even need to be super optimized and parralllel

**Reply by aikitoria** [reply]
> It'd be great if you could have CLIP and VAE loaded on one gpu and the unet on another card all in the same comfyui instance. Doesn't even need to be super optimized and parralllel

You can do this with https://github.com/neuratech-ai/ComfyUI-MultiGPU .

**Reply by Bratzmeister** [reply]
> > It'd be great if you could have CLIP and VAE loaded on one gpu and the unet on another card all in the same comfyui instance. Doesn't even need to be super optimized and parralllel
> 
> You can do this with https://github.com/neuratech-ai/ComfyUI-MultiGPU .

nice thx for pointing out! It should work well for flux and before but I'm afraid it's not suitable for newer models/architectures as it seems abandoned with last commit 5 months ago (3 month old reasonable PR ignored for tripleclip loader as well)

**Reply by aikitoria** [reply]
Could be. I've only used it for Flux. If the author abandoned it you could always fork it and pick it back up though.

**Reply by Bratzmeister** [reply]
true that but I still think this should be in core comfyui especially since more and more users tend to buy multiple cheaper gpus with newer generations not beefing up on VRAM much (if at all)

**Comment by tusharbhutt** [comment]
I'd just be happy if I can run multiple DIFFERENT prompts in parallel... seems a waste of a higher end GPU to be able to process 16 of the same prompt at once, or 16 (quickly) in succession, but not 16 different ones at once.  I know it can be done because I did it "hacking" a node (with zero coding skills) that made frame sequences, but that was under an older version of ComfyUI and it could not assign the prompt as filename correctly :(

**Reply by bagelbig** [reply]
I think you can spin up multiple instances of ComfyUI server on your machine, let's say 8, each with their own port.  Then open up a browser window to each instances (minding the different port #s).  You can then run multiple prompts (or indeed workflows) at the same time then.  I would recommend that each one is 'tweaked' to specify which GPU to run, or you can create a node that uses a flat file in the file system to check-in/check-out which GPU is being used, and then automatically determine which GPU is available and use that... could lead to race conditions, but those can be addressed through a more graceful implementation of what I suggested.

**Comment by WASasquatch** [comment]
Surprised we havent seen more from that concept of tiled rendering with
different noise/prompts for each tile to punch out a lot of unique content
at once.

**Comment by tempwswm** [comment]
https://github.com/comfyanonymous/ComfyUI/discussions/6647
how about this？

**Reply by vvauijij** [reply]
The Comfy frontend and Comfy backend seem to be tightly coupled, yet they serve different goals. I am currently researching ways to decouple them, enabling:
	1.	Scheduling different pipelines across multiple backends.
	2.	Executing different nodes of the same pipeline on multiple backends.

My motivation stems from the emergence of infrastructure-specific nodes (e.g., those improving VRAM caching, pinning models to specific GPUs, etc.). While these nodes enhance execution performance, placing them within original ML-specific pipelines feels misplaced. Instead, I aim to introduce greater execution flexibility, but the frontend and custom nodes doesn’t seem like the right place for this logic.

One approach could be integrating Ray as the backend backbone, bringing ComfyUI closer to a distributed execution model. However, I would consider:
	1.	Most pipelines used for personal tasks are relatively simple (loading weights, sampling, retrieving results) and might not benefit from the additional overhead of Ray’s abstraction.
	2.	Integrating Ray would turn the backend into a black box, limiting opportunities for configuration.
	3.	Configuring a Ray cluster with homogeneous nodes seems feasible, but it would require all nodes to have all necessary checkpoints, VAEs, etc., making dynamic node management (adding/removing nodes) less efficient.

**Reply by enn-nafnlaus** [reply]
> The Comfy frontend and Comfy backend seem to be tightly coupled, yet they serve different goals. I am currently researching ways to decouple them, enabling: 1. Scheduling different pipelines across multiple backends. 2. Executing different nodes of the same pipeline on multiple backends.
> 
> My motivation stems from the emergence of infrastructure-specific nodes (e.g., those improving VRAM caching, pinning models to specific GPUs, etc.). While these nodes enhance execution performance, placing them within original ML-specific pipelines feels misplaced. Instead, I aim to introduce greater execution flexibility, but the frontend and custom nodes doesn’t seem like the right place for this logic.
> 
> One approach could be integrating Ray as the backend backbone, bringing ComfyUI closer to a distributed execution model. However, I would consider: 1. Most pipelines used for personal tasks are relatively simple (loading weights, sampling, retrieving results) and might not benefit from the additional overhead of Ray’s abstraction. 2. Integrating Ray would turn the backend into a black box, limiting opportunities for configuration. 3. Configuring a Ray cluster with heterogeneous nodes seems feasible, but it would require all nodes to have all necessary checkpoints, VAEs, etc., making dynamic node management (adding/removing nodes) less efficient.

However it's done, I'd be appreciative. My current pipeline is to create "photographs" of fonts by making canny controlnet images of rendered fonts, in order to be able to use the "photos" to train a font-recognition ViT. Which means generating ideally millions of photos, since there's over 100k fonts. So I'm having to rent servers to do this, despite a very fast pipeline.  But my pipeline only spends like 70% of its time in the Ksampler, since again it's a very fast pipeline (only 4 steps, at 768x768 batch size 4); the rest is in image loading, image processing, and esp. output image saving.  So I'm only getting 70% utilization out of the GPUs I'm renting, so it's increasing my costs by over 40%.

**Reply by tempwswm** [reply]
1. It is true that simple, single-use workflows do not benefit from parallelism, but it is common that users will run the workflow multiple times to pick the better of the results. Also some common workflows may need to run batch processing such as processing all frames in a video.
2. ray providing a runtime environment does not limit its flexibility, most of the processing is still done by the node (actor).
3. machine synchronization is not a problem with Ray but with the cluster itself, instead it provides [shared memory](https://docs.ray.io/en/latest/ray-core/objects.html) and the [Ray Data](https://docs.ray.io/en/latest/data/data.html) library to improve these problems.

I don't think our discussion should focus on the need for parallelism, but rather how to achieve it in the least expensive way. Even parallel systems on the same machine have positive benefits.

**Reply by guill** [reply]
This is a super interesting proposal that I think is at least worth considering. Updating ComfyUI's execution model to handle parallel/asynchronous execution would be easy. Getting inference (and other torch operations that make up most of the execution time) to operate asynchronously is the difficult part. If Ray has already solved this challenge (and the resource management required for those operations), it may really accelerate ComfyUI's move to parallel execution.

**Reply by vvauijij** [reply]
I’ve been exploring the transition to parallel and distributed execution in ComfyUI, building on the original feature request by @tempwswm. After reviewing the ComfyUI source code and analyzing different systems, I’ve created a design doc that outlines how integrating Ray could enable better parallelism and scalability.

This is my first design doc, and I’d love to hear your feedback! Feel free to leave comments in the Notion doc if you have any suggestions or thoughts - https://www.notion.so/Introducing-Parallel-Distributed-Execution-for-ComfyUI-1a5d7138b2bf8028ab4dd35a4a5385b1?pvs=4

**Comment by mru4913** [comment]
How about serverless and multiple processes?

**Comment by monstari** [comment]
Any update on this? Is it possible to run single backend that uses both GPUs to process different workflows in parallel?

**Reply by aikitoria** [reply]
No, it's not possible yet.

**Comment by Dawgmastah** [comment]
Have there been any changes in how Comfyui manages multiple GPUs recently?
*Yes there were, sorry to bother*

My setup has an internal 3090ti, and an external (through thunderbolt) A6000
Normally I use either the MultiGPU nodes or set_cuda_visible_devices when launching Comfyui to force that instance/workflow to use one or the other. I has worked fine so far. It alllows me to run complicated WAN 2.1/2.2 or flux/context workflows saving lots of time in model loading.

However after the newest update, ComfyUi is defaulting to my internal 3090ti on default nodes no matter if its not even the first device listed in nvidia-smi or if I didnt even list it visible using set_cuda_visible_devices.

Is there a way I can return to the previous implementation?

*EDIT: This is a result of the default-device argument being introduced in 3.46, its more elegant than doing what I was doing.
